File: source/trees/avl/_static/code/avl.ts
import {
  assertEquals,
  assertThrows,
} from "https://deno.land/std@0.210.0/assert/mod.ts";

type Order<T> = (left: T, right: T) => boolean;

class OrderedSet<Item> {
  private _order: Order<Item>;
  private _root: Node<Item> | null;

  constructor(order: Order<Item>, root: Node<Item> | null) {
    this._order = order;
    this._root = root;
  }

  cardinal(): number {
    if (this.isEmpty) return 0;
    let cardinal = 0;
    const stack: Array<Node<Item>> = [this._root!];
    while (stack.length > 0) {
      const current = stack.pop();
      cardinal += 1;
      for (const eachChild of current!.children) {
        stack.push(eachChild);
      }
    }
    return cardinal;
  }

  get isEmpty() {
    return this._root == null;
  }

  contains(item: Item): boolean {
    if (this.isEmpty) return false;
    const path = this.findPathTo(item);
    return path[path.length - 1]!.item == item;
  }

  containsMany(...items: Array<Item>): boolean {
    for (const eachItem of items) {
      if (!this.contains(eachItem)) {
        return false;
      }
    }
    return true;
  }

  minimum(): Item {
    if (this.isEmpty) {
      throw new Error("Invalid state: An empty ordered set has no minimum.");
    }
    return this.minimumFrom(this._root!);
  }

  minimumFrom(root: Node<Item>) {
    let node = root;
    while (node != null && node.hasLeft) {
      node = node.left!;
    }
    return node!.item;
  }

  maximum(): Item {
    if (this.isEmpty) {
      throw new Error("Invalid state: An empty ordered set has no maximum.");
    }
    return this.maximumFrom(this._root!);
  }

  private maximumFrom(root: Node<Item>) {
    let node = root;
    while (node != null && node.hasRight) {
      node = node.right!;
    }
    return node!.item;
  }

  predecessorOf(item: Item): Item | undefined {
    if (this.isEmpty) {
      throw new Error(`Invalid state: Ordered set is empty.`);
    }
    const path = this.findPathTo(item);
    if (path[path.length - 1].item != item) {
      throw new Error(`Invalid state: Item '${item}' is not a member.`);
    } else {
      const node = path.pop();
      if (node!.hasLeft) {
        return this.maximumFrom(node!.left!);
      } else {
        return this.firstLesserAncestor(path, item);
      }
    }
  }

  private findPathTo(target: Item): Array<Node<Item>> {
    const path: Array<Node<Item>> = [];
    if (this.isEmpty) return path;
    let node = this._root;
    while (node != null) {
      path.push(node);
      if (this._order(node.item, target)) {
        if (node.item == target) {
          break;
        } else {
          node = node.right;
        }
      } else {
        node = node.left;
      }
    }
    return path;
  }

  private firstLesserAncestor(
    path: Array<Node<Item>>,
    item: Item,
  ): Item | undefined {
    while (path.length > 0) {
      const parent = path.pop();
      if (this._order(parent!.item, item)) {
        return parent!.item;
      }
    }
    return undefined;
  }

  successorOf(item: Item): Item | undefined {
    if (this.isEmpty) {
      throw new Error(`Invalid state: Ordered set is empty.`);
    }
    const path = this.findPathTo(item);
    if (path[path.length - 1].item != item) {
      throw new Error(`Invalid state: Item '${item}' is not a member.`);
    } else {
      const node = path.pop();
      if (node!.hasRight) {
        return this.minimumFrom(node!.right!);
      } else {
        return this.firstLargerAncestor(path, item);
      }
    }
  }

  private firstLargerAncestor(
    path: Array<Node<Item>>,
    item: Item,
  ): Item | undefined {
    while (path.length > 0) {
      const parent = path.pop();
      if (this._order(item, parent!.item)) {
        return parent!.item;
      }
    }
    return undefined;
  }

  add(newItem: Item) {
    if (this.isEmpty) {
      this._root = new Node<Item>(this._order, newItem, null, null);
    } else {
      const path = this.findPathTo(newItem);
      const parent = path[path.length - 1];
      const child = new Node<Item>(this._order, newItem, null, null);
      if (this._order(newItem, parent.item)) {
        if (parent.item != newItem) {
          parent.left = child;
        }
      } else {
        parent.right = child;
      }
      path.push(child);
      this.restoreBalance(path);
    }
  }

  restoreBalance(path: Array<Node<Item>>) {
    if (path.length < 3) return;
    let index = path.length - 3;
    while (index > 0) {
      if (path[index] == undefined) continue;
      const node = path[index];
      let update: Node<Item>;
      if (node.isRightHeavy()) {
        if (node.right.isRightHeavy()) {
          update = node.rotateLeft();
        } else {
          update = node.doubleRotateLeft();
        }
      } else if (node.isLeftHeavy()) {
        if (node.left.isLeftHeavy()) {
          update = node.rotateRight();
        } else {
          update = node.doubleRotateRight();
        }
      } else {
        update = node;
      }
      if (index == 0) {
        this._root = update;
      } else {
        this.implant(path[index - 1], update);
      }
      index--;
    }
  }

  implant(parent: Node<Item>, child: Node<Item>) {
    if (this._order(child.item, parent.item)) {
      parent.left = child;
    } else {
      parent.right = child;
    }
  }

  addMany(...items: Array<Item>) {
    for (const eachItem of items) {
      this.add(eachItem);
    }
  }

  remove(item: Item) {
    if (this.isEmpty) {
      throw new Error("Invalid state: Ordered set is empty.");
    } else {
      const path = this.findPathTo(item);
      const node = path.pop();
      if (node!.item != item) {
        throw new Error(`Invalid state: No item ${item}`);
      } else {
        const parent = path.pop();
        if (node!.children.length < 2) {
          if (parent) {
            parent.drop(node!);
          } else {
            this._root = node!.children[0];
          }
        } else {
          const predecessor = this.predecessorOf(node!.item);
          this.remove(predecessor!);
          node!.item = predecessor!;
        }
      }
    }
  }

  toString(): string {
    let result = "";
    if (this.isEmpty) return result;
    return this._root!.toString();
  }
}

class Node<Item> {
  private _order: Order<Item>;
  private _item: Item;
  private _left: Node<Item> | null;
  private _right: Node<Item> | null;

  constructor(
    order: Order<Item>,
    item: Item,
    left: Node<Item> | null,
    right: Node<Item> | null,
  ) {
    this._order = order;
    this._item = item;
    this._left = left;
    this._right = right;
  }

  get item() {
    return this._item;
  }

  set item(item: Item) {
    this._item = item;
  }

  set left(node: Node<Item> | null) {
    this._left = node;
  }

  get left(): Node<Item> {
    if (this._left == null) throw new Error("No left child");
    return this._left;
  }

  set right(node: Node<Item> | null) {
    this._right = node;
  }

  get right(): Node<Item> {
    if (this._right == null) throw new Error("No right child");
    return this._right;
  }

  get isLeaf(): boolean {
    return this._right == null && this._left == null;
  }

  drop(child: Node<Item>) {
    const descendant = child.isLeaf ? null : child.children[0];
    if (this._left == child) {
      this._left = descendant;
    }
    if (this._right == child) {
      this._right = descendant;
    }
  }

  // Bonjour les amis

  get children(): Array<Node<Item>> {
    const children: Array<Node<Item>> = [];
    if (this.hasLeft) children.push(this._left!);
    if (this.hasRight) children.push(this._right!);
    return children;
  }

  get hasRight(): boolean {
    return this._right != null;
  }

  get hasLeft(): boolean {
    return this._left != null;
  }

  balance() {
    const leftHeight = this.hasLeft ? this.left.height() : 0;
    const rightHeight = this.hasRight ? this.right.height() : 0;
    return leftHeight - rightHeight;
  }

  isBalanced() {
    const balance = this.balance();
    return balance >= -1 && balance <= 1;
  }

  isLeftHeavy() {
    return this.balance() > 0;
  }

  isRightHeavy() {
    return this.balance() < 0;
  }

  height() {
    const leftHeight = this.hasLeft ? this.left.height() : -1;
    const rightHeight = this.hasRight ? this.right.height() : -1;
    return 1 + Math.max(leftHeight, rightHeight);
  }

  rotateLeft(): Node<Item> {
    const oldRight = this.right;
    this.right = oldRight.left;
    oldRight.left = this;
    return oldRight;
  }

  rotateRight(): Node<Item> {
    const oldLeft = this.left;
    this.left = oldLeft.right;
    oldLeft.right = this;
    return oldLeft;
  }

  doubleRotateLeft(): Node<Item> {
    this.right = this.right.rotateRight();
    return this.rotateLeft();
  }

  doubleRotateRight(): Node<Item> {
    this.left = this.left.rotateLeft();
    return this.rotateRight();
  }

  toString(): string {
    if (this.isLeaf) return `${this.item}`;
    const left = this.hasLeft ? this._left!.toString() : "?";
    const right = this.hasRight ? this._right!.toString() : "?";
    return `(${this.item}: ${left}, ${right})`;
  }
}

Deno.test("cardinal is zero on create", () => {
  const tree = new OrderedSet<number>((l, r) => l <= r, null);
  assertEquals(tree.cardinal(), 0);
});

Deno.test("cardinal increases after add", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  assertEquals(oset.cardinal(), 0);
  oset.add(25);
  assertEquals(oset.cardinal(), 1);
});

Deno.test("cardinal does not increases when adding duplicate", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  assertEquals(oset.cardinal(), 0);
  oset.add(25);
  oset.add(25);
  assertEquals(oset.cardinal(), 1);
});

Deno.test("contains reveals added and missing items", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(25, 27, 12, 32, 12);

  assertEquals(oset.contains(25), true);
  assertEquals(oset.contains(27), true);
  assertEquals(oset.contains(32), true);
  assertEquals(oset.contains(12), true);

  assertEquals(oset.contains(13), false);
  assertEquals(oset.contains(26), false);
  assertEquals(oset.contains(0), false);
  assertEquals(oset.contains(999), false);
});

Deno.test("maximum returns the largest item", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertEquals(oset.maximum(), 58);
});

Deno.test("mminimum returns the smallest item", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertEquals(oset.minimum(), 12);
});

Deno.test("predecessor when the reference exists", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertEquals(oset.predecessorOf(23), 12);
  assertEquals(oset.predecessorOf(12), undefined);
  assertEquals(oset.predecessorOf(58), 28);
});

Deno.test("predecessor when the reference does not exists", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertThrows(() => oset.predecessorOf(0), Error);
  assertThrows(() => oset.predecessorOf(25), Error);
  assertThrows(() => oset.predecessorOf(99), Error);
});

Deno.test("successor when the reference exists", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertEquals(oset.successorOf(12), 23);
  assertEquals(oset.successorOf(23), 28);
  assertEquals(oset.successorOf(58), undefined);
});

Deno.test("successor when the reference does not exists", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertThrows(() => oset.successorOf(0), Error);
  assertThrows(() => oset.successorOf(25), Error);
  assertThrows(() => oset.successorOf(99), Error);
});

Deno.test("remove when the target does not exist", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);

  oset.addMany(23, 12, 58, 28);

  assertThrows(() => oset.remove(0), Error);
});

Deno.test("removed minimum are not available anymore", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  oset.addMany(23, 12, 58, 28);

  oset.remove(12);

  assertEquals(oset.contains(12), false);
  assertEquals(oset.containsMany(23, 28, 58), true);
});

Deno.test("removed root are not available anymore", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  oset.addMany(23, 12, 58, 28);

  oset.remove(23);

  assertEquals(oset.contains(23), false);
  assertEquals(oset.containsMany(12, 28, 58), true);
});

Deno.test("removed maximum is not available anymore", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  oset.addMany(23, 12, 58, 28);

  oset.remove(58);

  assertEquals(oset.contains(58), false);
  assertEquals(oset.containsMany(12, 23, 28), true);
});

Deno.test("remove singleton yields the empty set", () => {
  const oset = new OrderedSet<number>((l, r) => l <= r, null);
  oset.add(1);

  oset.remove(1);

  assertEquals(oset.contains(1), false);
  assertEquals(oset.cardinal(), 0);
});

File: source/trees/avl/_static/code/avl.py
class AVLTree:
    def __init__(self):
        self._size = 0
        self._root = None

    @property
    def balance_factor(self):
        if self.is_empty:
            return 0
        return self._root.balance_factor

    @property
    def is_empty(self):
        return self._size == 0

    @property
    def size(self) -> int:
        return self._size

    def insert(self, key, item):
        if self._root is None:
            self._root = AVLNode(Entry(key, item))
        else:
            self._root = self._root.insert(key, item)
        self._size += 1

    def get(self, key):
        if self._root is None:
            raise ValueError(f"No such key '{key}'")
        else:
            current_node = self._root
            while not current_node is None:
                if current_node.key == key:
                    return current_node.entry.item
                else:
                    current_node = current_node.branch_for(key)
            raise ValueError(f"No such key {key}")

    def predecessor(self, key):
        if self._root is not None:
            predecessor = self._root.predecessor(key)
            if predecessor is not None:
                return predecessor.entry.to_tuple
        return None

    def successor(self, key):
        if self._root is not None:
            successor = self._root.successor(key)
            if successor is not None:
                return successor.entry.to_tuple
        return None

    def contains(self, key):
        if self._root is not None:
            match = self._root.search(key)
            return match != None
        return False

    def delete(self, key):
        if self._root is None:
            raise ValueError(f"No such key {key}")
        self._root = self._root.delete(key)

    def items(self):
        if self._root is None:
            return
        current = self._root
        path = []
        while len(path) > 0 or current is not None:
            if current is not None:
                path.append(current)
                current = current.left if current.has_left else None
            else:
                current = path.pop()
                yield current.entry.item
                current = current.right if current.has_right else None

    def __repr__(self):
        return f"AVLTree({self._root})"


class Entry:
    def __init__(self, key, item):
        self._key = key
        self._item = item

    @property
    def key(self):
        return self._key

    @property
    def item(self):
        return self._item

    @property
    def to_tuple(self):
        return self._key, self._item

    def __repr__(self):
        return f"Entry({self._key}, {self._item})"


class AVLNode:
    def __init__(self, entry, left=None, right=None):
        if entry is None:
            raise ValueError("A node must have an entry")
        self._entry = entry
        self._left = left
        self._right = right

    @property
    def key(self):
        return self.entry.key

    @property
    def entry(self):
        return self._entry

    @property
    def has_left(self):
        return self._left is not None

    @property
    def left(self):
        if self._left is None:
            raise RuntimeError("Invalid state: No left branch")
        return self._left

    @left.setter
    def left(self, new_node):
        self._left = new_node

    @property
    def has_right(self):
        return self._right is not None

    @property
    def right(self):
        if self._right is None:
            raise RuntimeError(f"Invalid state: Node {self.key} has no right branch")
        return self._right

    @right.setter
    def right(self, new_node):
        self._right = new_node

    def branch_for(self, key):
        if key > self.key:
            return self._right
        else:
            return self._left

    def insert(self, key, item):
        if key == self.key:
            raise ValueError(f"Duplicate Key: {key}")
        elif key < self.key:
            if self.has_left:
                self.left = self.left.insert(key, item)
            else:
                self.left = AVLNode(Entry(key, item))
        else:
            if self.has_right:
                self.right = self.right.insert(key, item)
            else:
                self.right = AVLNode(Entry(key, item))
        return self.rebalance()

    @property
    def height(self):
        left_height = self._left.height if self.has_left else -1
        right_height = self._right.height if self.has_right else -1
        return 1 + max(left_height, right_height)

    @property
    def balance_factor(self):
        left_height = self._left.height if self.has_left else -1
        right_height = self._right.height if self.has_right else -1
        return left_height - right_height

    @property
    def is_balanced(self):
        return -2 < self.balance_factor < 2

    @property
    def is_unbalanced(self):
        return not self.is_balanced

    def rebalance(self):
        if self.is_balanced:
            return self
        if self.is_left_heavy and self.left.is_left_heavy:
            return self.rotate_right()
        elif self.is_left_heavy and self.left.is_right_heavy:
            self.left = self.left.rotate_left()
            return self.rotate_right()
        elif self.is_right_heavy and self.right.is_left_heavy:
            self.right = self.right.rotate_right()
            return self.rotate_left()
        else:
            return self.rotate_left()

    def rotate_right(self):
        new_root = self.left
        self.left = new_root.right if new_root.has_right else None
        new_root.right = self
        return new_root

    def rotate_left(self):
        new_root = self.right
        self.right = new_root.left if new_root.has_left else None
        new_root.left = self
        return new_root

    @property
    def is_right_heavy(self):
        return self.balance_factor < 0

    @property
    def is_left_heavy(self):
        return self.balance_factor > 0

    def predecessor(self, key):
        if self.key == key and self.has_left:
            return self.left.maximum
        elif key < self.key and self.has_left:
            return self.left.predecessor(key)
        elif key > self.key and self.has_right:
            return self.right.predecessor(key) or self
        else:
            return None

    def successor(self, key):
        if self.key == key and self.has_right:
            return self.right.minimum
        elif key < self.key and self.has_left:
            return self.left.successor(key) or self
        elif key > self.key and self.has_right:
            return self.right.successor(key)
        else:
            return None

    def __repr__(self):
        return f"AVLNode({self.entry}, {repr(self._left)}, {repr(self._right)})"

    @property
    def maximum(self):
        if self.has_right:
            return self.right.maximum
        else:
            return self

    @property
    def minimum(self):
        if self.has_left:
            return self.left.minimum
        else:
            return self

    def search(self, key):
        if self.key == key:
            return self
        elif key < self.key and self.has_left:
            return self.left.search(key)
        elif self.key < key and self.has_right:
            return self.right.search(key)
        else:
            return None

    def delete(self, key):
        if self.key == key:
            if self.has_left and self.has_right:
                predecessor = self.predecessor(key)
                self.delete(predecessor.key)
                predecessor.left = self.left if self.has_left else None
                predecessor.right = self.right if self.has_right else None
                return predecessor
            elif self.has_left:
                return self.left
            elif self.has_right:
                return self.right
            else:
                return None
        elif key < self.key and self.has_left:
            self.left = self.left.delete(key)
            return self.rebalance()
        elif self.key < key and self.has_right:
            self.right = self.right.delete(key)
            return self.rebalance()
        else:
            raise ValueError(f"No such key {key}")
